File: apps/ios/Packages/Data/Sources/Data/servers/ServersRepository.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import Common
import Model
import Network
import Persistence
internal import SeerrAPI

public protocol ServersRepository: Sendable {
    func observe() async -> AsyncThrowingStream<[ServerProfile], Error>
    func save(_ profile: ServerProfile) async throws
    func refresh(_ id: ProfileID) async throws
    func delete(_ id: ProfileID) async throws
}

struct LiveServersRepository: ServersRepository {
    let sessions: APISession
    let secrets: any SecretStore
    let preferences: PreferencesStore
    let users: UserCacheStore
    let titles: TitleCacheStore
    let clock: WallClock
    let images: ProfileImages

    func observe() async -> AsyncThrowingStream<[ServerProfile], Error> { await sessions.profiles.observe() }
    func save(_ profile: ServerProfile) async throws { try await sessions.profiles.save(profile) }

    func refresh(_ id: ProfileID) async throws {
        try await sessions.use(id) { profile, client, _ in
            let status: Operations.getStatus.Output.Ok.Body.jsonPayload
            switch try await client.getStatus(.init(query: .init(checkUpdateAvailable: false))) {
            case .ok(let response): status = try response.body.json
            case .undocumented(let code, _): throw responseError(code)
            }
            let settings: Components.Schemas.PublicSettings
            switch try await client.getSettingsPublic() {
            case .ok(let response): settings = try response.body.json
            case .undocumented(let code, _): throw responseError(code)
            }
            var updated = profile
            updated.status = Cached(
                ServerStatus(
                    version: status.version, commitTag: status.commitTag,
                    updateAvailable: status.updateAvailable, restartRequired: status.restartRequired),
                fetchedAt: clock.now())
            updated.publicSettings = Cached(
                PublicSettings(
                    title: settings.applicationTitle, initialized: settings.initialized,
                    localLogin: settings.localLogin,
                    mediaServerLogin: settings.mediaServerLogin,
                    mediaServerType: MediaServerType(rawValue: settings.mediaServerType.flatMap(Int.init(exactly:))),
                    cacheImages: settings.cacheImages, newPlexLogin: settings.newPlexLogin), fetchedAt: clock.now())
            try Task.checkCancellation()
            try await sessions.profiles.save(updated)
        }
    }

    func delete(_ id: ProfileID) async throws {
        try await sessions.transport.delete(id) {
            try await secrets.clear(profileID: id)
            try await users.clear(id)
            try await titles.clear(id)
            try await preferences.clear(id)
            try await images.clear(id)
            try await sessions.profiles.remove(id)
        }
    }
}

File: apps/android/core/data/src/main/kotlin/app/gauja/core/data/servers/ServersRepository.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.data.servers

import app.gauja.core.api.apis.PublicApi
import app.gauja.core.api.apis.SettingsApi
import app.gauja.core.data.session.ApiSession
import app.gauja.core.data.session.checked
import app.gauja.core.data.session.safeApi
import app.gauja.core.database.media.TitleDao
import app.gauja.core.database.users.UserDao
import app.gauja.core.datastore.preferences.PreferencesStore
import app.gauja.core.datastore.profiles.ServerProfileStore
import app.gauja.core.datastore.secrets.SecretStore
import app.gauja.core.model.servers.Cached
import app.gauja.core.model.servers.ProfileId
import app.gauja.core.model.servers.PublicSettings
import app.gauja.core.model.servers.ServerProfile
import app.gauja.core.model.servers.ServerStatus
import app.gauja.core.model.status.MediaServerType
import app.gauja.core.network.ProfileTransport
import java.time.Clock
import javax.inject.Inject
import kotlinx.collections.immutable.ImmutableList
import kotlinx.collections.immutable.toImmutableList
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

interface ServersRepository {
    val profiles: Flow<ImmutableList<ServerProfile>>

    suspend fun save(profile: ServerProfile)

    suspend fun refresh(profileId: ProfileId)

    suspend fun delete(profileId: ProfileId)
}

internal class LiveServersRepository
@Inject
constructor(
    private val store: ServerProfileStore,
    private val sessions: ApiSession,
    private val transport: ProfileTransport,
    private val secrets: SecretStore,
    private val preferences: PreferencesStore,
    private val users: UserDao,
    private val titles: TitleDao,
    private val clock: Clock,
) : ServersRepository {
    override val profiles = store.profiles.map { it.toImmutableList() }

    override suspend fun save(profile: ServerProfile) = safeApi { store.save(profile) }

    override suspend fun refresh(profileId: ProfileId) {
        sessions.use(profileId) { profile, retrofit ->
            val status = retrofit.create(PublicApi::class.java).getStatus(false).checked()
            val settings = retrofit.create(SettingsApi::class.java).getSettingsPublic().checked()
            val now = clock.instant()
            store.save(
                profile.copy(
                    status =
                        Cached(
                            ServerStatus(
                                status.version,
                                status.commitTag,
                                status.updateAvailable,
                                status.restartRequired,
                            ),
                            now,
                        ),
                    publicSettings =
                        Cached(
                            PublicSettings(
                                settings.applicationTitle,
                                settings.initialized,
                                settings.localLogin,
                                settings.mediaServerLogin,
                                MediaServerType.fromWire(
                                    settings.mediaServerType
                                        ?.takeIf { it.isFinite() && it % 1.0 == 0.0 }
                                        ?.toInt()
                                ),
                                settings.cacheImages,
                                newPlexLogin = settings.newPlexLogin,
                            ),
                            now,
                        ),
                )
            )
        }
    }

    override suspend fun delete(profileId: ProfileId) = safeApi {
        transport.delete(profileId) {
            secrets.clear(profileId)
            users.clear(profileId.value.toString())
            titles.clear(profileId.value.toString())
            preferences.clear(profileId)
            store.remove(profileId)
        }
    }
}
